cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,337 @@
+        boolean_expr: A Sympy expression containing symbols and Boolean operations
+        qubit_map: map of string (boolean variable name) to qubit.
+
+    Return:
+        The HamiltonianPolynomial that represents the Boolean expression.
+    """
+    if isinstance(boolean_expr, Symbol):
+        # Table 1 of [1], entry for 'x' is '1/2.I - 1/2.Z'
+        return 0.5 * _Hamiltonian_I() - 0.5 * _Hamiltonian_Z(qubit_map[boolean_expr.name])
+
+    if isinstance(boolean_expr, (And, Not, Or, Xor)):
+        sub_hamiltonians = [
+            _build_hamiltonian_from_boolean(sub_boolean_expr, qubit_map)
+            for sub_boolean_expr in boolean_expr.args
+        ]
+        # We apply the equalities of theorem 1 of [1].
+        if isinstance(boolean_expr, And):
+            hamiltonian = _Hamiltonian_I()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Not):
+            assert len(sub_hamiltonians) == 1
+            hamiltonian = _Hamiltonian_I() - sub_hamiltonians[0]
+        elif isinstance(boolean_expr, Or):
+            hamiltonian = _Hamiltonian_O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Xor):
+            hamiltonian = _Hamiltonian_O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - 2.0 * hamiltonian * sub_hamiltonian
+        return hamiltonian
+
+    raise ValueError(f'Unsupported type: {type(boolean_expr)}')
+
+
+def _gray_code_comparator(k1: Tuple[int, ...], k2: Tuple[int, ...], flip: bool = False) -> int:
+    """Compares two Gray-encoded binary numbers.
+
+    Args:
+        k1: A tuple of ints, representing the bits that are one. For example, 6 would be (1, 2).
+        k2: The second number, represented similarly as k1.
+        flip: Whether to flip the comparison.
+
+    Returns:
+        -1 if k1 < k2 (or +1 if flip is true)
+        0 if k1 == k2
+        +1 if k1 > k2 (or -1 if flip is true)
+    """
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def _get_gates_from_hamiltonians(
+    hamiltonian_polynomial_list: List['cirq.PauliSum'],
+    qubit_map: Dict[str, 'cirq.Qid'],
+    theta: float,
+    ladder_target: bool = False,
+):
+    """Builds a circuit according to [1].
+
+    Args:
+        hamiltonian_polynomial_list: the list of Hamiltonians, typically built by calling
+            _build_hamiltonian_from_boolean().
+        qubit_map: map of string (boolean variable name) to qubit.
+        theta: A single float scaling the rotations.
+        ladder_target: Whether to use convention of figure 7a or 7b.
```

Comment:
Can you elaborate on the difference between these two figures in words? It is hurting readability to have to read the paper back and forth.

Reply:
Please address this comment.

Reply:
Expanded the comment and added figures.